libcloudforensics/providers/kubernetes/base.py

```diff
@@ -0,0 +1,215 @@
+      T: An authenticated instance of the desired API class.
+    """
+    return api_class(self._api_client)
+
+
+class K8sResource(K8sClient, metaclass=abc.ABCMeta):
+  """Abstract class representing a Kubernetes resource."""
+
+  def __init__(self, api_client: client.ApiClient, name: str) -> None:
+    """Creates a Kubernetes resource holding Kubernetes API client.
+
+    Args:
+      api_client (ApiClient): The authenticated Kubernetes API client to
+        the cluster.
+      name (str): The name of this resource.
+    """
+    super().__init__(api_client)
+    self.name = name
+
+  @abc.abstractmethod
+  def Read(self) -> object:
+    """Returns the resulting read operation for the resource.
+
+    For example, a Node resource would call CoreV1Api.read_node, a Pod
+    resource would call CoreV1Api.read_namespaced_pod. The return values
+    of these read calls do not share a base class, hence the object return
+    type.
+
+    Returns:
+      object: The result of this resource's matching read operation.
+    """
+
+
+class K8sCluster(K8sClient):
+  """Class representing a Kubernetes cluster."""
+
+  def ListPods(self, namespace: Optional[str] = None) -> List['K8sPod']:
+    """Lists the pods of this cluster, possibly filtering for a namespace.
+
+    Args:
+      namespace (str): Optional. The namespace in which to list the pods.
+
+    Returns:
+      List[K8sPod]: The list of pods for the namespace, or in all namespaces
+        if none is specified.
+    """
+    api = self._Api(client.CoreV1Api)
+
+    # Collect pods
+    if namespace is not None:
+      pods = api.list_namespaced_pod(namespace)
+    else:
+      pods = api.list_pod_for_all_namespaces()
+
+    # Convert to node objects
+    return [K8sPod(self._api_client, pod.metadata.name, pod.metadata.namespace)
+            for pod in pods.items]
+
+  def ListNodes(self) -> List['K8sNode']:
+    """Lists the nodes of this cluster.
+
+    Returns:
+      List[K8sNode]: The list of nodes in this cluster.
+    """
+    api = self._Api(client.CoreV1Api)
+
+    # Collect pods
+    nodes = api.list_node()
+
+    # Convert to node objects
+    return [K8sNode(self._api_client, node.metadata.name)
+            for node in nodes.items]
+
+
+class K8sNamespacedResource(K8sResource, metaclass=abc.ABCMeta):
+  """Class representing a Kubernetes resource, in a certain namespace."""
+
+  def __init__(self, api_client: client.ApiClient, name: str,
+               namespace: str) -> None:
+    """Creates a Kubernetes resource in the given namespace.
+
+    Args:
+      api_client (ApiClient): The authenticated Kubernetes API client to
+        the cluster.
+      name (str): The name of this resource.
+      namespace (str): The Kubernetes namespace in which this resource
+        resides
+    """
+    super().__init__(api_client, name)
+    self.namespace = namespace
+
+
+class K8sNode(K8sResource):
+  """Class representing a Kubernetes node."""
+
+  def Read(self) -> client.V1Node:
```

Comment:
Docstring?

Reply:
Added brief docstring as per Google's style guide: https://google.github.io/styleguide/pyguide.html#383-functions-and-methods
> A method that overrides a method from a base class may have a simple docstring sending the reader to its overridden method’s docstring, such as `"""See base class."""`.